### DISCUSSION TITLE
Removable (Remote) Media

### DISCUSSION THREAD

**Discussion by oliverschmidt** [tjboldt/Apple2-IO-RPi#107] [General]
Hi,

So far, Apple2-IO-RPi clearly emulates two fixed hard disks. Bute there are scenarios, where one would rather prefer to emulate some removable media. That would require two enhancements from my pov:
1. Being able to close and reopen the two disk image files during a2driver runtime.
2. A user interface for setting/choosing/selecting the pathname of the two disk image files - ideally from the Apple II.

I consider this relevant, because one of the benefits of using a Linux micro-computer over using some micro-controller is the flexibility coming with remote file systems. Without further code changes in Apple2-IO-RPi, the Apple II can boot from / access all types of locations by just providing a pathname to a2driver that that points to a remote file system.

I have two specific use cases in mind:
1. [cifs-utils](https://wiki.samba.org/index.php/LinuxCIFS_utils) allows to mount any SMB file system into the local file system. Many users own some NAS (maybe just embedded in some other device). Directly using disk images located there would e.g. allow to manipulate them via CiderPress without having to move anything back and forth.
2. [httpfs2](http://httpfs.sourceforge.net/) allows to mount a HTTP(S) URL into the local file system. This is _NOT_ WebDAV, this is just using [byte serving](https://en.wikipedia.org/wiki/Byte_serving) which is supported by about every HTTP server. Using the Total Replay URL would e.g. allow to play those games right from the Internet.

Beside the two enhancements mentioned above, using remote file systems makes an additional enhancement desirable:

Being able to deal with errors while accessing disk image files. The current Apple2-IO-RPi protocol doesn't allow a2driver to communicate any error back to ProDOS. You simply can't have e.g. an `IO ERROR`, rather it just hangs until you reset. But when you place a disk image file on some remote file system, it can easily happen that the file can't be accessed at some point - and that should be handles in a reasonable way. A specific aspect is a disk image file that is read-only which is of course always the case for httpfs2. The Apple II should be able to get a `WRITE PROTECTED` error back from a2driver.

Regards, Oliver

**Comment by Hubidrei** [comment]
I like the Idea of removeable, exchangeable drives.
Like it is possible on WINAPPLE. 
My Idea is to have a small lcd on the pi to choose the images to be mounted.
I have done so with a STM32Bluepill, small ssd1306 oled, 3 keys, CH367 . I can mount .po, .dsk, .hdv and access the files. No Bootrom yet. Serial communication. Stopped for a while, or forever, since i changed to Apple2-io-rpi.
